File: CST ASP.NET @ PCC for MDAC/ADO.NET Course Files/Labs/Lab05/Solution/Ex3/CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.IO;
using System.Xml;

namespace LoadingDataSets
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class frmCustomer : System.Windows.Forms.Form
	{
		private const string myDocument = @"\Program Files\MSDNTrain\2389\Labs\Lab05\customers.xml";
		private const string myLoadSchema = @"\Program Files\MSDNTrain\2389\Labs\Lab05\customerschema.xsd";
		private const string mySaveSchema = @"\Program Files\MSDNTrain\2389\Labs\Lab05\ResultSchema.xsd";
		private const string mySaveDocument = @"\Program Files\MSDNTrain\2389\Labs\Lab05\ResultData.xml";
		private const string mySaveBoth = @"\Program Files\MSDNTrain\2389\Labs\Lab05\ResultInlineSchema.xml";
		private DataSet myDS;
		
		private System.Windows.Forms.DataGrid dataGrid1;
		private System.Windows.Forms.Button btnLoadData;
		private System.Windows.Forms.Button btnClose;
		private System.Windows.Forms.CheckBox chkInferSchema;
		private System.Windows.Forms.Button btnSaveSchema;
		private System.Windows.Forms.Button btnSaveData;
		private System.Windows.Forms.Button btnSaveBoth;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public frmCustomer()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.btnClose = new System.Windows.Forms.Button();
			this.chkInferSchema = new System.Windows.Forms.CheckBox();
			this.dataGrid1 = new System.Windows.Forms.DataGrid();
			this.btnLoadData = new System.Windows.Forms.Button();
			this.btnSaveSchema = new System.Windows.Forms.Button();
			this.btnSaveData = new System.Windows.Forms.Button();
			this.btnSaveBoth = new System.Windows.Forms.Button();
			((System.ComponentModel.ISupportInitialize)(this.dataGrid1)).BeginInit();
			this.SuspendLayout();
			// 
			// btnClose
			// 
			this.btnClose.Location = new System.Drawing.Point(232, 240);
			this.btnClose.Name = "btnClose";
			this.btnClose.Size = new System.Drawing.Size(168, 24);
			this.btnClose.TabIndex = 2;
			this.btnClose.Text = "Close";
			this.btnClose.Click += new System.EventHandler(this.btnClose_Click);
			// 
			// chkInferSchema
			// 
			this.chkInferSchema.Location = new System.Drawing.Point(232, 40);
			this.chkInferSchema.Name = "chkInferSchema";
			this.chkInferSchema.TabIndex = 3;
			this.chkInferSchema.Text = "Infer Schema";
			// 
			// dataGrid1
			// 
			this.dataGrid1.DataMember = "";
			this.dataGrid1.Dock = System.Windows.Forms.DockStyle.Left;
			this.dataGrid1.Name = "dataGrid1";
			this.dataGrid1.Size = new System.Drawing.Size(224, 273);
			this.dataGrid1.TabIndex = 0;
			// 
			// btnLoadData
			// 
			this.btnLoadData.Location = new System.Drawing.Point(232, 8);
			this.btnLoadData.Name = "btnLoadData";
			this.btnLoadData.Size = new System.Drawing.Size(168, 23);
			this.btnLoadData.TabIndex = 1;
			this.btnLoadData.Text = "Display Customer Information";
			this.btnLoadData.Click += new System.EventHandler(this.btnLoadData_Click);
			// 
			// btnSaveSchema
			// 
			this.btnSaveSchema.Location = new System.Drawing.Point(232, 80);
			this.btnSaveSchema.Name = "btnSaveSchema";
			this.btnSaveSchema.Size = new System.Drawing.Size(168, 23);
			this.btnSaveSchema.TabIndex = 4;
			this.btnSaveSchema.Text = "Save Schema";
			this.btnSaveSchema.Click += new System.EventHandler(this.btnSaveSchema_Click);
			// 
			// btnSaveData
			// 
			this.btnSaveData.Location = new System.Drawing.Point(232, 112);
			this.btnSaveData.Name = "btnSaveData";
			this.btnSaveData.Size = new System.Drawing.Size(168, 23);
			this.btnSaveData.TabIndex = 5;
			this.btnSaveData.Text = "Save Data";
			this.btnSaveData.Click += new System.EventHandler(this.btnSaveData_Click);
			// 
			// btnSaveBoth
			// 
			this.btnSaveBoth.Location = new System.Drawing.Point(232, 144);
			this.btnSaveBoth.Name = "btnSaveBoth";
			this.btnSaveBoth.Size = new System.Drawing.Size(168, 23);
			this.btnSaveBoth.TabIndex = 6;
			this.btnSaveBoth.Text = "Save Both";
			this.btnSaveBoth.Click += new System.EventHandler(this.btnSaveBoth_Click);
			// 
			// frmCustomer
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(408, 273);
			this.Controls.AddRange(new System.Windows.Forms.Control[] {
																		  this.btnSaveBoth,
																		  this.btnSaveData,
																		  this.btnSaveSchema,
																		  this.chkInferSchema,
																		  this.btnClose,
																		  this.btnLoadData,
																		  this.dataGrid1});
			this.Name = "frmCustomer";
			this.Text = "Sales Information";
			((System.ComponentModel.ISupportInitialize)(this.dataGrid1)).EndInit();
			this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new frmCustomer());
		}

		private void btnClose_Click(object sender, System.EventArgs e)
		{
			this.Close();
		}

		private void btnLoadData_Click(object sender, System.EventArgs e)
		{

			try
			{
				myDS = new DataSet();

				if(this.chkInferSchema.Checked)
				{
					myDS.ReadXml(myDocument, XmlReadMode.InferSchema);
				}
				else
				{
					myDS.ReadXmlSchema(myLoadSchema);
					myDS.ReadXml(myDocument, XmlReadMode.IgnoreSchema);
				}

				this.dataGrid1.DataSource = myDS.Tables["Customers"];
			}
			catch (Exception xcp)
			{
				MessageBox.Show(xcp.ToString());
			}

		}

		private void btnSaveSchema_Click(object sender, System.EventArgs e)
		{
			myDS.WriteXmlSchema(mySaveSchema);
		}

		private void btnSaveData_Click(object sender, System.EventArgs e)
		{
			myDS.WriteXml(mySaveDocument, XmlWriteMode.IgnoreSchema);
		}

		private void btnSaveBoth_Click(object sender, System.EventArgs e)
		{
			myDS.WriteXml(mySaveBoth, XmlWriteMode.WriteSchema);
		}
	}
}

File: CST ASP.NET @ PCC for MDAC/ADO.NET Course Files/Labs/Lab04_2/Solution/Ex2/VB/Form1.vb
Public Class Form1
    Inherits System.Windows.Forms.Form

    ' Added in Lab 4.1 Exercise 2

    Public Filename As System.String

    ' Windows Form Designer generated variables

    Friend WithEvents grd As System.Windows.Forms.DataGrid
    Friend WithEvents dlgOpen As System.Windows.Forms.OpenFileDialog
    Friend WithEvents dlgSave As System.Windows.Forms.SaveFileDialog


    Friend WithEvents mnuFile As System.Windows.Forms.MenuItem
    Friend WithEvents mnuNew As System.Windows.Forms.MenuItem
    Friend WithEvents mnuOpen As System.Windows.Forms.MenuItem
    Friend WithEvents s1 As System.Windows.Forms.MenuItem
    Friend WithEvents mnuSave As System.Windows.Forms.MenuItem
    Friend WithEvents mnuSaveAs As System.Windows.Forms.MenuItem
    Friend WithEvents s2 As System.Windows.Forms.MenuItem
    Friend WithEvents mnuExit As System.Windows.Forms.MenuItem

    Friend WithEvents mnuView As System.Windows.Forms.MenuItem
    Friend WithEvents mnuCustomer As System.Windows.Forms.MenuItem
    Friend WithEvents mnuProducts As System.Windows.Forms.MenuItem
    Friend WithEvents mnuMain As System.Windows.Forms.MainMenu
    Friend WithEvents dsShoppingCart As System.Data.DataSet
    Friend WithEvents dtCustomer As System.Data.DataTable
    Friend WithEvents dtCartItems As System.Data.DataTable
    Friend WithEvents dcCustomerID As System.Data.DataColumn
    Friend WithEvents dcCompanyName As System.Data.DataColumn
    Friend WithEvents dcAddress As System.Data.DataColumn
    Friend WithEvents dcCity As System.Data.DataColumn
    Friend WithEvents dcCustomerID2 As System.Data.DataColumn
    Friend WithEvents dcProductID As System.Data.DataColumn
    Friend WithEvents dcUnitPrice As System.Data.DataColumn
    Friend WithEvents dcQuantity As System.Data.DataColumn
    Friend WithEvents dcCost As System.Data.DataColumn
    Friend WithEvents mnuSubtotal As System.Windows.Forms.MenuItem
    Friend WithEvents MenuItem2 As System.Windows.Forms.MenuItem
        Friend WithEvents mnuCartItems As System.Windows.Forms.MenuItem

#Region " Windows Form Designer generated code "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()

        'Add any initialization after the InitializeComponent() call

    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub

    'Required by the Windows Form Designer
    Private components As System.ComponentModel.Container

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
        Me.mnuExit = New System.Windows.Forms.MenuItem()
        Me.mnuNew = New System.Windows.Forms.MenuItem()
        Me.dcCost = New System.Data.DataColumn()
        Me.dcAddress = New System.Data.DataColumn()
        Me.s2 = New System.Windows.Forms.MenuItem()
        Me.mnuCustomer = New System.Windows.Forms.MenuItem()
        Me.dcProductID = New System.Data.DataColumn()
        Me.mnuOpen = New System.Windows.Forms.MenuItem()
        Me.mnuSave = New System.Windows.Forms.MenuItem()
        Me.mnuProducts = New System.Windows.Forms.MenuItem()
        Me.grd = New System.Windows.Forms.DataGrid()
        Me.mnuCartItems = New System.Windows.Forms.MenuItem()
        Me.dcCity = New System.Data.DataColumn()
        Me.dtCustomer = New System.Data.DataTable()
        Me.dcCustomerID = New System.Data.DataColumn()
        Me.dcCompanyName = New System.Data.DataColumn()
        Me.mnuSaveAs = New System.Windows.Forms.MenuItem()
        Me.dsShoppingCart = New System.Data.DataSet()
        Me.dtCartItems = New System.Data.DataTable()
        Me.dcCustomerID2 = New System.Data.DataColumn()
        Me.dcUnitPrice = New System.Data.DataColumn()
        Me.dcQuantity = New System.Data.DataColumn()
        Me.dlgSave = New System.Windows.Forms.SaveFileDialog()
        Me.mnuFile = New System.Windows.Forms.MenuItem()
        Me.s1 = New System.Windows.Forms.MenuItem()
        Me.mnuSubtotal = New System.Windows.Forms.MenuItem()
        Me.dlgOpen = New System.Windows.Forms.OpenFileDialog()
        Me.mnuMain = New System.Windows.Forms.MainMenu()
        Me.mnuView = New System.Windows.Forms.MenuItem()
        Me.MenuItem2 = New System.Windows.Forms.MenuItem()
        CType(Me.grd, System.ComponentModel.ISupportInitialize).BeginInit()
        CType(Me.dtCustomer, System.ComponentModel.ISupportInitialize).BeginInit()
        CType(Me.dsShoppingCart, System.ComponentModel.ISupportInitialize).BeginInit()
        CType(Me.dtCartItems, System.ComponentModel.ISupportInitialize).BeginInit()
        Me.SuspendLayout()
        '
        'mnuExit
        '
        Me.mnuExit.Index = 6
        Me.mnuExit.Text = "E&xit"
        '
        'mnuNew
        '
        Me.mnuNew.Index = 0
        Me.mnuNew.Text = "&New"
        '
        'dcCost
        '
        Me.dcCost.ColumnName = "Cost"
        Me.dcCost.DataType = GetType(System.Decimal)
        Me.dcCost.Expression = "UnitPrice * Quantity"
        Me.dcCost.ReadOnly = True
        '
        'dcAddress
        '
        Me.dcAddress.ColumnName = "Address"
        Me.dcAddress.MaxLength = 60
        '
        's2
        '
        Me.s2.Index = 5
        Me.s2.Text = "-"
        '
        'mnuCustomer
        '
        Me.mnuCustomer.Index = 0
        Me.mnuCustomer.Shortcut = System.Windows.Forms.Shortcut.F7
        Me.mnuCustomer.Text = "&Customer"
        '
        'dcProductID
        '
        Me.dcProductID.AllowDBNull = False
        Me.dcProductID.ColumnName = "ProductID"
        Me.dcProductID.DataType = GetType(System.Int32)
        '
        'mnuOpen
        '
        Me.mnuOpen.Index = 1
        Me.mnuOpen.Text = "&Open..."
        '
        'mnuSave
        '
        Me.mnuSave.Index = 3
        Me.mnuSave.Shortcut = System.Windows.Forms.Shortcut.CtrlS
        Me.mnuSave.Text = "&Save"
        '
        'mnuProducts
        '
        Me.mnuProducts.Index = 2
        Me.mnuProducts.Shortcut = System.Windows.Forms.Shortcut.CtrlF7
        Me.mnuProducts.Text = "&Products"
        '
        'grd
        '
        Me.grd.DataMember = ""
        Me.grd.Dock = System.Windows.Forms.DockStyle.Fill
        Me.grd.Name = "grd"
        Me.grd.Size = New System.Drawing.Size(368, 289)
        Me.grd.TabIndex = 0
        '
        'mnuCartItems
        '
        Me.mnuCartItems.Index = 1
        Me.mnuCartItems.Shortcut = System.Windows.Forms.Shortcut.ShiftF7
        Me.mnuCartItems.Text = "Cart &Items"
        '
        'dcCity
        '
        Me.dcCity.ColumnName = "City"
        Me.dcCity.MaxLength = 15
        '
        'dtCustomer
        '
        Me.dtCustomer.Columns.AddRange(New System.Data.DataColumn() {Me.dcCustomerID, Me.dcCompanyName, Me.dcAddress, Me.dcCity})
        Me.dtCustomer.TableName = "Customer"
        '
        'dcCustomerID
        '
        Me.dcCustomerID.AllowDBNull = False
        Me.dcCustomerID.ColumnName = "CustomerID"
        Me.dcCustomerID.MaxLength = 5
        '
        'dcCompanyName
        '
        Me.dcCompanyName.AllowDBNull = False
        Me.dcCompanyName.ColumnName = "CompanyName"
        Me.dcCompanyName.MaxLength = 40
        '
        'mnuSaveAs
        '
        Me.mnuSaveAs.Index = 4
        Me.mnuSaveAs.Text = "Save &As..."
        '
        'dsShoppingCart
        '
        Me.dsShoppingCart.DataSetName = "NewDataSet"
        Me.dsShoppingCart.Locale = New System.Globalization.CultureInfo("en-US")
        Me.dsShoppingCart.Tables.AddRange(New System.Data.DataTable() {Me.dtCustomer, Me.dtCartItems})
        '
        'dtCartItems
        '
        Me.dtCartItems.Columns.AddRange(New System.Data.DataColumn() {Me.dcCustomerID2, Me.dcProductID, Me.dcUnitPrice, Me.dcQuantity, Me.dcCost})
        Me.dtCartItems.TableName = "CartItems"
        '
        'dcCustomerID2
        '
        Me.dcCustomerID2.AllowDBNull = False
        Me.dcCustomerID2.ColumnName = "CustomerID"
        Me.dcCustomerID2.MaxLength = 5
        '
        'dcUnitPrice
        '
        Me.dcUnitPrice.AllowDBNull = False
        Me.dcUnitPrice.ColumnName = "UnitPrice"
        Me.dcUnitPrice.DataType = GetType(System.Decimal)
        '
        'dcQuantity
        '
        Me.dcQuantity.AllowDBNull = False
        Me.dcQuantity.ColumnName = "Quantity"
        Me.dcQuantity.DataType = GetType(System.Int32)
        '
        'dlgSave
        '
        Me.dlgSave.DefaultExt = "ds"
        Me.dlgSave.Filter = "DataSet files (*.ds)|*.ds"
        '
        'mnuFile
        '
        Me.mnuFile.Index = 0
        Me.mnuFile.MenuItems.AddRange(New System.Windows.Forms.MenuItem() {Me.mnuNew, Me.mnuOpen, Me.s1, Me.mnuSave, Me.mnuSaveAs, Me.s2, Me.mnuExit})
        Me.mnuFile.Text = "&File"
        '
        's1
        '
        Me.s1.Index = 2
        Me.s1.Text = "-"
        '
        'mnuSubtotal
        '
        Me.mnuSubtotal.Index = 4
        Me.mnuSubtotal.Shortcut = System.Windows.Forms.Shortcut.F9
        Me.mnuSubtotal.Text = "Cart &Subtotal"
        '
        'dlgOpen
        '
        Me.dlgOpen.Filter = "DataSet files (*.ds)|*.ds|All files (*.*)|*.*"
        '
        'mnuMain
        '
        Me.mnuMain.MenuItems.AddRange(New System.Windows.Forms.MenuItem() {Me.mnuFile, Me.mnuView})
        '
        'mnuView
        '
        Me.mnuView.Index = 1
        Me.mnuView.MenuItems.AddRange(New System.Windows.Forms.MenuItem() {Me.mnuCustomer, Me.mnuCartItems, Me.mnuProducts, Me.MenuItem2, Me.mnuSubtotal})
        Me.mnuView.Text = "&View"
        '
        'MenuItem2
        '
        Me.MenuItem2.Index = 3
        Me.MenuItem2.Text = "-"
        '
        'Form1
        '
        Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
        Me.ClientSize = New System.Drawing.Size(368, 289)
        Me.Controls.AddRange(New System.Windows.Forms.Control() {Me.grd})
        Me.Menu = Me.mnuMain
        Me.Name = "Form1"
        Me.Text = "Form1"
        CType(Me.grd, System.ComponentModel.ISupportInitialize).EndInit()
        CType(Me.dtCustomer, System.ComponentModel.ISupportInitialize).EndInit()
        CType(Me.dsShoppingCart, System.ComponentModel.ISupportInitialize).EndInit()
        CType(Me.dtCartItems, System.ComponentModel.ISupportInitialize).EndInit()
        Me.ResumeLayout(False)

    End Sub

#End Region

    ' Added in Lab 4.1 Exercise 2

    Public Sub OpenFromFile()

        Try

            ' clear any existing data
            Me.dsShoppingCart.Tables("CartItems").Clear()
            Me.dsShoppingCart.Tables("Customer").Clear()

            ' open the file
            Me.dsShoppingCart.ReadXml(Me.Filename)

        Catch Xcp As System.Exception

            MessageBox.Show(Xcp.ToString(), _
                "Failed to open: " & Me.Filename, _
                MessageBoxButtons.OK, MessageBoxIcon.Exclamation)

        End Try

    End Sub

    ' Added in Lab 4.1 Exercise 2

    Public Sub SaveToFile()

        Try

            ' save the file
            Me.dsShoppingCart.WriteXml(Me.Filename)

        Catch Xcp As System.Exception

            MessageBox.Show(Xcp.ToString(), _
                "Failed to save: " & Me.Filename, _
                MessageBoxButtons.OK, MessageBoxIcon.Exclamation)

        End Try

    End Sub

    ' Added in Lab 4.1 Exercise 3

    Private Sub SetFormCaption()

        Me.Text = Me.Filename & " - Shopping Cart : Test WinApp"

    End Sub

    Private Sub mnuCustomer_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuCustomer.Click

        ' Added in Lab 4.1 Exercise 3

        Me.grd.DataSource = New DataView(Me.dsShoppingCart.Tables("Customer"))

        Me.mnuCustomer.Checked = True
        Me.mnuCartItems.Checked = False

    End Sub

    Private Sub mnuCartItems_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuCartItems.Click

        ' Added in Lab 4.1 Exercise 3

        Me.grd.DataSource = New DataView(Me.dsShoppingCart.Tables("CartItems"))

        Me.mnuCustomer.Checked = False
        Me.mnuCartItems.Checked = True

    End Sub

    Private Sub mnuExit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuExit.Click

        With Me
            .Dispose()
            .Close()
        End With

    End Sub

    Private Sub mnuNew_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuNew.Click

        ' clear any existing data
        Me.dsShoppingCart.Tables("CartItems").Clear()
        Me.dsShoppingCart.Tables("Customer").Clear()

        Me.Filename = "ShoppingCart1.ds"

        Me.SetFormCaption()

        Me.mnuCustomer_Click(sender, e)

    End Sub

    Private Sub mnuOpen_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles mnuOpen.Click

        If Me.dlgOpen.ShowDialog() = DialogResult.OK Then

            Me.Filename = Me.dlgOpen.FileName

            Me.OpenFromFile()

            Me.SetFormCaption()

            Me.mnuCustomer_Click(sender, e)

        End If

    End Sub

    Private Sub mnuSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuSave.Click

        Me.SaveToFile()

    End Sub

    Private Sub mnuSaveAs_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuSaveAs.Click

        If Me.dlgSave.ShowDialog() = DialogResult.OK Then

            Me.Filename = Me.dlgSave.FileName

            Me.SaveToFile()

            Me.SetFormCaption()

        End If

    End Sub

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Me.mnuNew_Click(sender, e)

        ' add primary key

        Me.dtCustomer.Constraints.Add("PK_Customer", _
            Me.dtCustomer.Columns("CustomerID"), True)

        ' add relation

        Me.dsShoppingCart.Relations.Add("FK_Customer_CartItems", _
            Me.dtCustomer.Columns("CustomerID"), _
            Me.dtCartItems.Columns("CustomerID"), True)

        ' add subtotal column

        Me.dtCustomer.Columns.Add(New DataColumn( _
            "CartSubtotal", GetType(System.Decimal), _
            "Sum(Child.Cost)"))

    End Sub

    Private Sub mnuSubtotal_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuSubtotal.Click

        Dim sCustomerID As System.String
        Dim drCustomer As System.Data.DataRow
        Dim drCartItem As System.Data.DataRow
        Dim dSubtotal As System.Decimal = 0

        Try

            sCustomerID = Me.grd(Me.grd.CurrentRowIndex, 0)

            drCustomer = Me.dsShoppingCart.Tables("Customer").Select( _
                "CustomerID='" & sCustomerID & "'")(0)

            For Each drCartItem In drCustomer.GetChildRows("FK_Customer_CartItems")

                dSubtotal += drCartItem.Item("Cost")

            Next

            MessageBox.Show(drCustomer("CompanyName") & " owes " & dSubtotal.ToString("$0.00"), "Cart Subtotal")

        Catch Xcp As System.Exception

            MessageBox.Show(Xcp.ToString, "Unexpected exception", _
                MessageBoxButtons.OK, MessageBoxIcon.Exclamation)

        End Try

    End Sub

End Class
